pr for review

sankethkini opened the pull request.

sankethkini committed:
application part, added validators and declared required function

sankethkini committed:
added required constants

sankethkini committed:
models user and course is added

sankethkini committed:
added file adapters for saving files in disk

changed rollnumber field to string

sankethkini committed:
added more testcase for fileadapters

sankethkini committed:
added memory adapter for storing data in memory

memory adapters takes data from application if it's array than it just sorts whole array
if it is a single entry it will sort it using insertion sort in o(n)

sudiksa17 commented on application/application.go:
```diff
@@ -0,0 +1,49 @@
+	curroll := userdata["rollnum"].(string)
+	if len(curroll) < 2 {
+		return NoRollNum
+	}
```
give space after one fucntionality
```
curname := userdata["fname"].(string)
	if len(curname) < 2 {
		return NoNameErr
	}

curroll := userdata["rollnum"].(string)
	if len(curroll) < 2 {
		return NoRollNum
	}
```

sudiksa17 commented on adapters/memoryAdapter.go:
```diff
@@ -0,0 +1,74 @@
+		adapter.Items = append(adapter.Items, cur)
+		last := len(adapter.Items) - 1
+		for last > 0 {
+			if adapter.Items[last-1].Fname < adapter.Items[last].Fname {
```
revisit this, too much nesting

sankethkini committed:
added ui components

UI components for displaying menu and other option

sankethkini committed:
main file of the program

sankethkini committed:
modified application file added more functionality

added display,add, delete and save functionality
also tested with appropriate test cases

sankethkini committed:
modified adapters added remaining functionality

adapter init for singleton implementation
retriveall functionality implemented with different sorting

sudiksa17 commented on UI/cli.go:
```diff
@@ -0,0 +1,218 @@
+package UI
```
package name: lowercase
Go through Effective Go doc: https://go.dev/doc/effective_go#package-names

sudiksa17 commented on UI/cli.go:
```diff
@@ -0,0 +1,218 @@
+}
+
+//cli func for display users
+func displayUser() {
```
a single function should do a single task

sudiksa17 commented on main.go:
```diff
@@ -0,0 +1,7 @@
+import "github.com/sankethkini/StudentData/UI"
+
+func main() {
+	UI.Start()
```
this function returns error, why its not checked here
run linter

sudiksa17 commented on adapters/memoryAdapter.go:
```diff
@@ -0,0 +1,179 @@
+
+	switch field {
+	case "name":
+		if order == 1 {
```
duplicate code in all switch cases

sudiksa17 commented:
run linter
go through golang docs for the functionality that u r using : effective go doc and golang style guide
a single function should do a single task

sankethkini committed:
seperating file and memory adapters using interfaces in a better way

sankethkini committed:
seperating diffrent functions for differnt jobs used global adapter

sankethkini committed:
seperating functions for diff jobs and taking user input using buffer

sankethkini committed:
removed unwanted files

sankethkini committed:
removed unwanted files

sankethkini committed:
small changes in tests

sudiksa17 commented on .vscode/launch.json:
```diff
@@ -0,0 +1,15 @@
+{
```
add `.gitignore`

sudiksa17 commented on ui/cli.go:
```diff
@@ -0,0 +1,235 @@
+
+//cli for exit
+func exit() {
+	fmt.Println("exiting ..... ")
```
does the exit functionality asks user whether they want to save the changes

sudiksa17 commented on adapter/memory/errors.go:
```diff
@@ -0,0 +1,6 @@
+import "errors"
+
+var RecordNotFound = errors.New("record not found")
+var NotARightType = errors.New("not a right type")
```
why defining same error var again

sudiksa17 commented on application/application.go:
```diff
@@ -0,0 +1,195 @@
+func Add(userdata data) ([]data, error) {
+
+	//checking validity of user input
+	validationerr := inputValidator(userdata)
```
var name: camelCase `validationErr`, check other places as well

sudiksa17 commented on domain/user/user.go:
```diff
@@ -0,0 +1,11 @@
+	Age     int
+	Adress  string
+	RollNo  string
+	Courses []course.Course
```
can define `Course` struct in user package

sudiksa17 commented on adapter/file/file.go:
```diff
@@ -0,0 +1,113 @@
+		fileinfo.Close()
+	}
+
+	fileinfo, err := os.Open(constants.Filelocation)
```
length of the name of var should depend on the scope of variable use

sudiksa17 commented on adapter/memory/memory.go:
```diff
@@ -0,0 +1,180 @@
+
+//RetriveAll function to retrive all records
+//sorting based on field
+func (adpt *adapter) RetriveAll(field string, order int) (interface{}, error) {
```
why does this func returns generic interface{} if its implemented only once and returns only single type of struct []User

sankethkini committed:
changed interfaces, used proper conventions to name error

sankethkini committed:
added constructor to user struct

sankethkini committed:
added .gitignore

sankethkini committed:
some minor change according to linter

sankethkini committed:
deleted unwanted files

sankethkini committed:
removed unwanted files

sudiksa17 commented on ui/cli.go:
```diff
@@ -0,0 +1,238 @@
+	} else if ch == "n" {
+		fmt.Println("exiting program without saving user")
+	} else {
+		exit()
```
never ending loop if user selects anything other than y / n

sudiksa17 commented on ui/cli.go:
```diff
@@ -0,0 +1,238 @@
+		data["field"] = "address"
+	default:
+		data["field"] = "name"
+		fmt.Println("Enter the valid number")
```
we are not prompting user for any input, what purpose will this logger serve

sudiksa17 commented on ui/cli.go:
```diff
@@ -0,0 +1,238 @@
+	//taking course input from user
+	selectedCourse := takeCourseInput()
+
+	userdata := make(map[string]interface{})
```
`userdata` to `userData`

sudiksa17 commented on ui/cli.go:
```diff
@@ -0,0 +1,238 @@
+	userdata["rollnum"] = rollnum
+	userdata["address"] = address
+	userdata["age"] = age
+	userdata["courses"] = selectedCourse
```
instead of having map, can we directly set through NewUser

sudiksa17 commented on application/application_test.go:
```diff
@@ -0,0 +1,188 @@
+package application
```
package test failing
increase coverage

sankethkini committed:
added validators in user domain file, used ozzo validation

sankethkini committed:
changed code according to review

sankethkini committed:
added makefile

sudiksa17 commented on go.mod:
```diff
@@ -0,0 +1,3 @@
+module github.com/sankethkini/StudentData
```
run `go mod tidy` and read about why go `mod` is used and what `go mod tidy` do

sankethkini replied:
done

sankethkini committed:
applied go mod tidy

sankethkini committed:
changes done according to linter and previous reviews

sankethkini committed:
changes in test cases
